src/imitation/algorithms/preference_comparisons.py

```diff
@@ -641,50 +680,59 @@ def __call__(
-                var_estimates.append(rew1_var + rew2_var)
-            else:  # uncertainty_on is probability or label
-                with th.no_grad():
-                    rews1 = self.preference_predictor.rewards(trans1)
-                    rews2 = self.preference_predictor.rewards(trans2)
-                    probs = (
-                        self.preference_predictor.probability(rews1, rews2)
-                        .cpu()
-                        .numpy()
-                    )
-                if self.uncertainty_on == "probability":
-                    var_estimates.append(probs.var())
-                else:  # uncertainty_on is label
-                    preds = (probs > 0.5).astype(np.float32)
-                    # probability estimate of Bernoulli random variable
-                    prob_estimate = preds.mean()
-                    # variance estimate of Bernoulli random variable
-                    var_estimate = prob_estimate * (1 - prob_estimate)
-                    var_estimates.append(var_estimate)
-
+            with th.no_grad():
+                rews1 = self.preference_model.rewards(trans1)
+                rews2 = self.preference_model.rewards(trans2)
+            var_estimate = self.variance_estimate(rews1, rews2)
+            var_estimates[i] = var_estimate
         fragment_idxs = np.argsort(var_estimates)[::-1]  # sort in descending order
         # return fragment pairs that have the highest uncertainty
         return [fragment_pairs[idx] for idx in fragment_idxs[:num_pairs]]
 
-    def _reward_moments(self, transitions):
-        rewards = self.preference_predictor.rewards(transitions).sum(0)
-        return rewards.mean().cpu().numpy(), rewards.var().cpu().numpy()
+    def variance_estimate(self, rews1, rews2) -> float:
+        """Gets the variance estimate from the rewards of a fragment pair.
+
+        Args:
+            rews1: rewards obtained by all the ensemble models for the first fragment.
+                Shape - (fragment_length, num_ensemble_members)
+            rews2: rewards obtained by all the ensemble models for the second fragment.
+                Shape - (fragment_length, num_ensemble_members)
+
+        Returns:
+            the variance estimate based on the `uncertainty_on` flag.
+        """
+        if self.uncertainty_on == "logit":
+            rews1, rews2 = rews1.sum(0), rews2.sum(0)
+            var_estimate = (rews1 - rews2).var().item()
+        else:  # uncertainty_on is probability or label
+            probs = (
+                self.preference_model.probability(
+                    rews1,
+                    rews2,
+                )
+                .cpu()
+                .numpy()
+            )
+            if self.uncertainty_on == "probability":
+                var_estimate = probs.var()
+            else:  # uncertainty_on is label
+                preds = (probs > 0.5).astype(np.float32)
+                # probability estimate of Bernoulli random variable
```

Comment:
I'm treating it as a Bernoulli random variable since the random variable we are capturing is the preferred fragment (between fragment0 and fragment1). So if 60% of the ensemble member networks prefer fragment1 over fragment0 (i.e., they assign `prob` > 0.5), then `preds` will be an array of 60% 1s and 40% 0s. Finally, `prob_estimate` would be the Bernoulli probability estimate of preferring fragment1 over fragment0 which will be 60%. 

Does this make sense, or am I confusing/mixing things up?

Reply:
This makes sense, thanks for explanation, I was thinking of the Bernoulli random variable as being based on `probs` rather than `preds` being an estimate of a latent Bernoulli that the ensemble as a whole is modeling. Thanks for clearing this up!